Add support for Rust

Hi, I'd like to build code for ESP32 in Rust, on my NixOS machine. I though this would be a good start. Unfortunately, this does not work yet. When I tried the following:

```console
https://github.com/espressif/rust-esp32-example.git
cd rust-esp32-example
idf.py set-target esp32
idf.py build
```
which errors with 

```console
  process didn't exit successfully: `rustc - --crate-name ___ --print=file-names --target xtensa-esp32-espidf --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=cfg` (exit status: 1)
  --- stderr
  error: Error loading target specification: Could not find specification for target "xtensa-esp32-espidf". Run `rustc --print target-list` for a list of built-in targets
```

This is expected - I did not install the Rust fork yet. However, and this is where I think we might have a problem in the nixpkgs-esp-dev project:

```console
https://github.com/espressif/rust-esp32-example.git
cd rust-esp32-example
idf.py set-target esp32c3
idf.py build
```

results in 

```console
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
-- The ASM compiler identification is unknown
-- Found assembler: riscv32-esp-elf-gcc
CMake Error at /nix/store/zxyj5pafs2v9dbc1ij30zxg00cyi20a9-esp-idf-v4.3.1/tools/cmake/project.cmake:311 (__project):
  The CMAKE_C_COMPILER:

    riscv32-esp-elf-gcc

  is not a full path and was not found in the PATH.

  Tell CMake where to find the compiler by setting either the environment
  variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
  the compiler, or to the compiler name if it is in the PATH.
Call Stack (most recent call first):
  CMakeLists.txt:6 (project)


CMake Error at /nix/store/zxyj5pafs2v9dbc1ij30zxg00cyi20a9-esp-idf-v4.3.1/tools/cmake/project.cmake:311 (__project):
  The CMAKE_CXX_COMPILER:

    riscv32-esp-elf-g++

  is not a full path and was not found in the PATH.

  Tell CMake where to find the compiler by setting either the environment
  variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
  to the compiler, or to the compiler name if it is in the PATH.
Call Stack (most recent call first):
  CMakeLists.txt:6 (project)


CMake Error at /nix/store/zxyj5pafs2v9dbc1ij30zxg00cyi20a9-esp-idf-v4.3.1/tools/cmake/project.cmake:311 (__project):
  The CMAKE_ASM_COMPILER:

    riscv32-esp-elf-gcc

  is not a full path and was not found in the PATH.

  Tell CMake where to find the compiler by setting either the environment
  variable "ASM" or the CMake cache entry CMAKE_ASM_COMPILER to the full path
  to the compiler, or to the compiler name if it is in the PATH.
Call Stack (most recent call first):
  CMakeLists.txt:6 (project)


-- Warning: Did not find file Compiler/-ASM
-- Configuring incomplete, errors occurred!
```

Is this nix version of the toolchain supposed to already work with the esp32c3 (which needs a RISC-V toolchain)?

--- a/README.md
+++ b/README.md
@@ -1,5 +1,5 @@
 # nixpkgs-esp-dev
-ESP8266 and ESP32 packages and development environments for Nix.
+ESP8266 and ESP32(-C3) packages and development environments for Nix.
 
 This repo contains derivations for:
 - Toolchains (compiler, linker, GDB, etc.) for `xtensa-lx106-elf` (ESP8266) and `xtensa-esp32-elf` (ESP32) using the official binaries from Espressif.
@@ -21,6 +21,7 @@ The easiest way to get started is to run one of these commands to get a developm
 ### `nix-shell`
 If you're not using Nix 2.4+ or prefer not to need to enable flakes, you can clone the repo and use one of:
 
+- `nix-shell shells/esp32c3-idf.nix`
 - `nix-shell shells/esp32-idf.nix`
 - `nix-shell shells/esp8266.nix`
 

--- a/flake.nix
+++ b/flake.nix
@@ -15,7 +15,7 @@
     {
       packages = {
         inherit (pkgs)
-          gccriscv32-esp32c3-elf-bin
+          gcc-riscv32-esp32c3-elf-bin
           gcc-xtensa-esp32-elf-bin
           openocd-esp32-bin
           esp-idf

--- a/flake.nix
+++ b/flake.nix
@@ -15,6 +15,7 @@
     {
       packages = {
         inherit (pkgs)
+          gccriscv32-esp32c3-elf-bin
           gcc-xtensa-esp32-elf-bin
           openocd-esp32-bin
           esp-idf
@@ -25,6 +26,7 @@
       };
 
       devShells = {
+        esp32c3-idf = import ./shells/esp32c3-idf.nix { inherit pkgs; };
         esp32-idf = import ./shells/esp32-idf.nix { inherit pkgs; };
         esp8266 = import ./shells/esp8266.nix { inherit pkgs; };
       };

--- a/overlay.nix
+++ b/overlay.nix
@@ -15,6 +15,8 @@ let
   };
 in
 {
+  # ESP32C3
+  gcc-riscv32-esp32c3-elf-bin = prev.callPackage ./pkgs/esp32c3-toolchain-bin.nix { };
   # ESP32
   gcc-xtensa-esp32-elf-bin = prev.callPackage ./pkgs/esp32-toolchain-bin.nix { };
   openocd-esp32-bin = prev.callPackage ./pkgs/openocd-esp32-bin.nix { };

--- a/pkgs/esp32c3-toolchain-bin.nix
+++ b/pkgs/esp32c3-toolchain-bin.nix
@@ -1,0 +1,52 @@
+# This version needs to be compatible with the version of ESP-IDF specified in `esp-idf/default.nix`.
+{ version ? "2021r1"
+, hash ? "sha256-NFlhjzO71fVNfXeD6AfLbu9kcqIg8vHrP6ztc1udE7s="
+, stdenv
+, lib
+, fetchurl
+, makeWrapper
+, buildFHSUserEnv
+}:
+
+let
+  fhsEnv = buildFHSUserEnv {
+    name = "esp32c3-toolchain-env";
+    targetPkgs = pkgs: with pkgs; [ zlib ];
+    runScript = "";
+  };
+in
+
+assert stdenv.system == "x86_64-linux";
+
+stdenv.mkDerivation rec {
+  pname = "esp32c3-toolchain";
+  inherit version;
+
+  src = fetchurl {
+    url = "https://github.com/espressif/crosstool-NG/releases/download/esp-${version}/riscv32-esp-elf-gcc8_4_0-esp-${version}-linux-amd64.tar.gz";
+    inherit hash;
+  };
+
+  buildInputs = [ makeWrapper ];
+
+  phases = [ "unpackPhase" "installPhase" ];
+
+  installPhase = ''
+    cp -r . $out
+    for FILE in $(ls $out/bin); do
+      FILE_PATH="$out/bin/$FILE"
+      echo "$FILE_PATH"
+      if [[ -x $FILE_PATH ]]; then
+        mv $FILE_PATH $FILE_PATH-unwrapped
+        makeWrapper ${fhsEnv}/bin/esp32c3-toolchain-env $FILE_PATH --add-flags "$FILE_PATH-unwrapped"
+      fi
+    done
+  '';
+
+  meta = with lib; {
+    description = "ESP32-C3 compiler toolchain";
+    homepage = "https://docs.espressif.com/projects/esp-idf/en/stable/get-started/linux-setup.html";
+    license = licenses.gpl3;
+  };
+}
+

--- a/shells/esp32c3-idf.nix
+++ b/shells/esp32c3-idf.nix
@@ -1,0 +1,27 @@
+{ pkgs ? import ../default.nix }:
+
+pkgs.mkShell {
+  name = "esp-idf";
+
+  buildInputs = with pkgs; [
+    gcc-riscv32-esp32c3-elf-bin
+    openocd-esp32-bin
+    esp-idf
+    esptool
+
+    # Tools required to use ESP-IDF.
+    git
+    wget
+    gnumake
+
+    flex
+    bison
+    gperf
+    pkgconfig
+
+    cmake
+    ninja
+
+    ncurses5
+  ];
+}

